src/Microsoft.OData.Core/UriParser/Parsers/UriQueryExpressionParser.cs

```diff
@@ -205,6 +205,38 @@ internal static string GetEdmConstantNames(IEdmTypeReference edmTypeReference)
             }
         }
 
+        // parses $compute query option.
+        internal IEnumerable<ComputeToken> ParseCompute(string compute)
+        {
+            Debug.Assert(compute != null, "compute != null");
+
+            List<ComputeToken> transformationTokens = new List<ComputeToken>();
+
+            if (string.IsNullOrEmpty(compute))
+            {
+                return transformationTokens;
+            }
+
+            this.recursionDepth = 0;
+            this.lexer = CreateLexerForFilterOrOrderByOrApplyExpression(compute);
+
+            while (true)
+            {
+                ComputeToken computed = this.ParseComputeExpression();

```

Comment:
computed can't be null since we are testing for the a null compute string and returning empty token list if encountered. computed can be invalid in a variety of ways. your negative testing comment points in that direction. invalid expression formation is caught be the reuse of the expression binding code path, only the 'as' tests should be improved.  testing the alias that follows the 'as' is already covered by the apply code path that is being reused.

---
In reply to: [122828867](https://github.com/OData/odata.net/pull/827#discussion_r122828867) [](ancestors = 122828867)